### PR TITLE
win,pipe: support vectored writes, fix IPC deadlock

### PR DESCRIPTION
**Pull request by piscisaureus**
Fixes: https://github.com/libuv/libuv/issues/794
Fixes: https://github.com/libuv/libuv/issues/1099
Refs: https://github.com/nodejs/node/issues/7657
Refs: https://github.com/electron/electron/issues/10107
Refs: https://github.com/parcel-bundler/parcel/issues/637
Refs: https://github.com/parcel-bundler/parcel/issues/900
Refs: https://github.com/parcel-bundler/parcel/issues/1137

### CODE REVIEW

File: src/win/pipe.c [hunk 1]
```diff
@@ -1656,106 +1744,34 @@ void uv_process_pipe_read_req(uv_loop_t* loop, uv_pipe_t* handle,
...
-        }
-        handle->read_cb((uv_stream_t*)handle, bytes, &buf);
 
-        /* Read again only if bytes == buf.len */
-        if (bytes <= buf.len) {
-          break;
-        }
-      } else {
-        uv_pipe_read_error_or_eof(loop, handle, GetLastError(), buf);
+      /* It is possible that more bytes more bytes were read than we thought
```
Comment (bzoz) [inline src/win/pipe.c]: "more bytes more bytes"

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/847/
CI-node: https://ci.nodejs.org/view/libuv/job/libuv-in-node/20/

File: src/win/pipe.c (hunk 1 above)
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

**Comment by bzoz** [comment]
CI is red, on Linux it complains about undefined reference to `ipc_helper_heavy_traffic_deadlock_bug`, on Windows one of the test is failing:
```console
not ok 150 - pipe_getsockname_blocking
# exit code -1073741819
# Output from process `pipe_getsockname_blocking`: (no output)
```

**Comment by piscisaureus** [comment]
@bnoordhuis 
I am at a loss about why the test added in be7978d2c276821f09148807f3f783ecf2d9291b fails on linux.
It seems that writev() blocks after writing ~160kb of data to the socket.
I would not expect it to do that; it is properly set to nonblocking mode, so why doesn't it fail with EAGAIN?

**Comment by santigimeno** [comment]
> It seems that writev() blocks after writing ~160kb of data to the socket.

I haven't analyzed the test but by calling the test with `strace` it does not seem that way: it seems every call to `writev` completes. These are the last lines before the test runner kills the test due to exceed the timeout

```
[pid  4225] 12:42:38.858250 <... writev resumed> ) = 24576
[pid  4223] 12:42:38.858257 <... epoll_wait resumed> [{EPOLLIN, {u32=11, u64=11}}], 1024, -1) = 1
[pid  4225] 12:42:38.858267 writev(0, [{iov_base="********************************"..., iov_len=8192}, {iov_base="********************************"..., iov_len=8192}, {iov_base="********************************"..., iov_len=8192}], 3 <unfinished ...>
[pid  4223] 12:42:38.858282 recvmsg(11,  <unfinished ...>
[pid  4225] 12:42:38.858293 <... writev resumed> ) = 24576
[pid  4223] 12:42:38.858306 <... recvmsg resumed> {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="********************************"..., iov_len=65536}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_CMSG_CLOEXEC) = 49152
[pid  4225] 12:42:38.858318 brk(0x556880b48000 <unfinished ...>
[pid  4223] 12:42:38.858353 epoll_wait(4,  <unfinished ...>
[pid  4225] 12:42:38.858751 <... brk resumed> ) = 0x556880b48000
[pid  4225] 12:42:38.858766 epoll_wait(4, [{EPOLLOUT, {u32=0, u64=0}}], 1024, 0) = 1
[pid  4225] 12:42:38.858792 epoll_ctl(4, EPOLL_CTL_DEL, 0, 0x7ffe5796edb0) = 0
[pid  4225] 12:42:38.858822 epoll_wait(4,
```

**Comment by piscisaureus** [comment]
@santigimeno If you run the test without supervision of the test harness (`out/Debug/run-tests ipc_heavy_traffic_deadlock_bug ipc_heavy_traffic_deadlock_bug`), does it complete for you on linux?

**Comment by gireeshpunathil** [comment]
I see two potential issues:
 - `ASSERT(nread >= 0);` in `read_cb` : I guess we cannot guarantee this given the uv_read can return negative values on error?
 - it is so possible that the test takes longer than the prescribed timeout by the runner?

File: src/win/pipe.c [hunk 2]
```diff
@@ -1422,7 +1349,7 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
       req->u.io.queued_bytes = 0;
     } else {
       /* Request queued by the kernel. */
-      req->u.io.queued_bytes = bufs[0].len;
+      req->u.io.queued_bytes = write_buf.len;
       handle->write_queue_size += req->u.io.queued_bytes;
       if (WaitForSingleObject(req->u.io.overlapped.hEvent, INFINITE) !=
           WAIT_OBJECT_0) {
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: Yeah, that looks incorrect. The code path is likely never taken, I don't think `WaitForSingleObject` can conceivably fail, unless hEvent is invalid, in which case there is a bug elsewhere.

File: test/test-ipc-heavy-traffic-deadlock-bug.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply test/test-ipc-heavy-traffic-deadlock-bug.c]: Right, thanks, copypasta.

**Comment by santigimeno** [comment]
What I see is that the child process receives all the data and exits correctly but the parent process is stuck. It looks some pending handle needs to be closed.

**Comment by piscisaureus** [comment]
@gireeshpunathil 

> ASSERT(nread >= 0); in read_cb : I guess we cannot guarantee this given the uv_read can return negative values on error?

Where in the code is this / should it be?

> It is so possible that the test takes longer than the prescribed timeout by the runner?

I think that's unlikely. The tests makes 256 write calls and sends a grand total of 6mb of data to the child process. The child process does the same in the other direction. That should not even take a second.

File: test/test-ipc-heavy-traffic-deadlock-bug.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                     uv_buf_t* buf) {
+  buf->base = malloc(suggested_size);
+  buf->len = (int) suggested_size;
+}
+
+static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
+  ssize_t i;
+  int r;
+
+  ASSERT(nread >= 0);
```
Comment (gireeshpunathil) [inline test/test-ipc-heavy-traffic-deadlock-bug.c]: I believe this is overloaded with the number of bytes written | error code from the syscall

**Comment by gireeshpunathil** [comment]
@piscisaureus - highlighted the point above. Agreed on the timeout stuff, shouldn't take that much ( I was speculating). Now that @santigimeno has spotted a hang, I guess that looks to be the right direction.

**Comment by santigimeno** [comment]
The ipc channel needs to be closed. This patch fixes the issue locally for me:
```c
diff --git a/test/test-ipc-heavy-traffic-deadlock-bug.c b/test/test-ipc-heavy-traffic-deadlock-bug.c
index 62868e6a..d1b54bff 100644
--- a/test/test-ipc-heavy-traffic-deadlock-bug.c
+++ b/test/test-ipc-heavy-traffic-deadlock-bug.c
@@ -81,6 +81,15 @@ static void alloc_cb(uv_handle_t* handle,
 static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
   ssize_t i;
   int r;
+  uv_shutdown_t shut_req;
+
+  if (nread == UV_EOF) {
+    ASSERT(bytes_read == XFER_SIZE);
+    r = uv_read_stop(handle);
+    ASSERT(r == 0);
+    uv_close((uv_handle_t*)handle, NULL);
+    return;
+  }
 
   ASSERT(nread >= 0);
   bytes_read += nread;
@@ -91,7 +100,7 @@ static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
   free(buf->base);
 
   if (bytes_read >= XFER_SIZE) {
-    r = uv_read_stop(handle);
+    uv_shutdown(&shut_req, handle, NULL);
     ASSERT(r == 0);
   }
 }
```

**Comment by piscisaureus** [comment]
libuv ci looks green: https://ci.nodejs.org/job/libuv-test-commit/857/

**Comment by piscisaureus** [comment]
libuv-in-node ci: https://ci.nodejs.org/view/libuv/job/libuv-in-node/

File: src/win/pipe.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

File: src/win/pipe.c [hunk 5]
```diff
@@ -1478,28 +1429,139 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
...
+  int err;
+
+  /* Compute the combined size of data buffers. */
+  data_length = 0;
+  for (i = 0; i < data_buf_count; i++) {
+    data_length += data_bufs[i].len;
+  }
+  if (data_length > UINT32_MAX) {
+    return WSAENOBUFS; /* Maps to UV_ENOBUFS. */
+  }
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

File: src/win/pipe.c [hunk 6]
```diff
@@ -1226,154 +1216,115 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
...
+
+  /* Compute heap buffer size. */
+  heap_buffer_length = sizeof *coalesced_write_req + /* (a) + (b) */
+                       data_length;                  /* (c) */
+
+  /* Allocate buffer. */
+  heap_buffer = uv__malloc(heap_buffer_length);
+  if (heap_buffer == NULL) {
+    return ERROR_NOT_ENOUGH_MEMORY; /* Maps to UV_ENOMEM. */
+  }
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

File: src/win/pipe.c [hunk 7]
```diff
@@ -1226,154 +1216,115 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
...
+  data_length = 0;
+  for (i = 0; i < nbufs; i++) {
+    data_length += bufs[i].len;
+  }
+
+  /* The total combined size of data buffers should not exceed UINT32_MAX,
+   * because WriteFile() won't accept buffers larger than that. */
+  if (data_length > UINT32_MAX) {
+    return WSAENOBUFS; /* Maps to UV_ENOBUFS. */
+  }
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

File: src/win/pipe.c [hunk 8]
```diff
@@ -1226,154 +1216,115 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
...
+  uv__coalesced_write_t* coalesced_write_req; /* (a) + (b) */
+  char* data_start;                           /* (c) */
+  size_t data_length;
+  unsigned int i;
+
+  /* Compute combined size of all combined buffers from `bufs`. */
+  data_length = 0;
+  for (i = 0; i < nbufs; i++) {
+    data_length += bufs[i].len;
+  }
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/win/pipe.c]: fixed

File: test/test.gyp [hunk 9]
```diff
@@ -48,6 +48,7 @@
         'test-idle.c',
         'test-ip6-addr.c',
         'test-ipc.c',
+        'test-ipc-heavy-traffic-deadlock-bug.c',
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply test/test.gyp]: fixed

File: test/test-ipc-heavy-traffic-deadlock-bug.c [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+  r = uv_run(handle->loop, UV_RUN_DEFAULT);
+  ASSERT(r == 0);
+
+  ASSERT(bytes_written == XFER_SIZE);
+  ASSERT(bytes_read == XFER_SIZE);
+}
+
+TEST_IMPL(ipc_heavy_traffic_deadlock_bug) {
+  uv_loop_t* loop = uv_default_loop();
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply test/test-ipc-heavy-traffic-deadlock-bug.c]: `loop` was unused anyway, so I removed it.

File: test/test-ipc-heavy-traffic-deadlock-bug.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (piscisaureus) [reply test/test-ipc-heavy-traffic-deadlock-bug.c]: fixed

**Comment by piscisaureus** [comment]
CI:
https://ci.nodejs.org/job/libuv-test-commit/859/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/24

In node, there seems to be one occasional failure that is related to this patch series: `test-child-process-fork-net` sometimes dies due to an unhandled EPIPE error.
This is expected - the tests spawns a child process and attempts to send a message to it after the child process has exited, so the pipe between them breaks. In other words, the test is buggy.

I would suggest to land this and fix the test in node.

@bzoz, @santigimeno, @gireeshpunathil PTAL

**Comment by bzoz** [comment]
Node test is broken right now: https://github.com/nodejs/node/issues/20836

But this looks related (from https://ci.nodejs.org/job/node-test-binary-windows/17578/COMPILED_BY=vs2017,RUNNER=win10,RUN_SUBSET=0/console):
```console
not ok 7 async-hooks/test-graph.pipeconnect
  ---
  duration_ms: 0.311
  severity: fail
  exitcode: -1073740791
  stack: |-
    Assertion failed: r || GetLastError() == ERROR_NOT_FOUND, file src\win\pipe.c, line 708
  ...
```

**Comment by piscisaureus** [comment]
https://ci.nodejs.org/job/libuv-test-commit/860/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/25/

**Comment by piscisaureus** [comment]
https://ci.nodejs.org/job/libuv-test-commit/862/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/26/console

**Comment by piscisaureus** [comment]
https://ci.nodejs.org/view/libuv/job/libuv-test-commit/863/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/27/

**Comment by piscisaureus** [comment]
also
https://ci.nodejs.org/view/libuv/job/libuv-test-commit/864/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/28/ (with node v10.x)

**Comment by piscisaureus** [comment]
https://ci.nodejs.org/view/libuv/job/libuv-test-commit/866/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/28/ (with node/v10.x)
https://ci.nodejs.org/view/libuv/job/libuv-in-node/27/ (with node/master)
^-- pretty green, no failures related to this patch series.

@bzoz, @santigimeno PTAL

**Comment by gireeshpunathil** [comment]
LGTM

**Comment by piscisaureus** [comment]
I think this is done. 
Can I get one more approval from @bzoz, @santigimeno, @cjihrig, @bnoordhuis.
(I understand the patch set is a large -- but it also solves a real problem and I would rather not have it bitrot away.)

File: src/win/pipe.c [hunk 11]
```diff
@@ -1226,154 +1227,111 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
...
+           bufs[i].len);               /* copy (c) */
+    heap_buffer_offset += bufs[i].len; /* offset (c) */
+  }
+  assert(heap_buffer_offset == heap_buffer_length);
+
+  /* Set out arguments and return. */
+  *req_out = &coalesced_write_req->req;
+  *write_buf_out = uv_buf_init(data_start, (unsigned int) data_length);
+  return 0;
+}
```
Comment (addaleax) [inline src/win/pipe.c]: Just curious, but did you look into strong a list of the user-supplised `uv_buf_t`s instead? It seems like that might be a slightly more direct approach. (Even if you think that would work, I’m no asking you to put the work in of changing it, it’s really just a question.)

Comment (piscisaureus) [reply src/win/pipe.c]: @addaleax I agree, it is the more obvious approach. The reason it doesn't work is that there is no syscall that writes to pipes and takes a vector of buffers. 
An alternative approach might have been to write the first buffer, wait for the completion port to report that the write was successful, then write the second, etc. But there are serious drawbacks as well:
* It involves making more syscalls, so it's probably slower (unless the messages sent over the IPC pipe are very big).
* State management gets a lot more complex. And pipe.c is already a complex beast :(
* The requirement to not change the ABI makes that I can't the add necessary extra fields to uv_write_t.

**Comment by piscisaureus** [comment]
@addaleax Thanks for your review!

@ all: I am going to keep this PR open until Tuesday morning PST to allow for the other reviewers to raise further objections or approve.

File: src/win/pipe.c [hunk 12]
```diff
@@ -1226,154 +1227,111 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
...
-                              const uv_buf_t bufs[],
-                              unsigned int nbufs,
-                              uv_stream_t* send_handle,
-                              uv_write_cb cb) {
-  int err;
-  int result;
-  uv_tcp_t* tcp_send_handle;
-  uv_write_t* ipc_header_req = NULL;
-  uv_ipc_frame_uv_stream ipc_frame;
+static int uv__build_coalesced_write_req(uv_write_t* user_req,
```
Comment (bzoz) [inline src/win/pipe.c]: I think we should mention this in the docs, that the buffers will get combined

Comment (piscisaureus) [reply src/win/pipe.c]: I'm not sure I agree, as this entirely transparent (and likely unnoticeable) to the user.
For now, created an issue for it: https://github.com/libuv/libuv/issues/1855

**Pull request merged**